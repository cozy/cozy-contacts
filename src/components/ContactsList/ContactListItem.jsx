import React, { Component } from 'react'
import PropTypes from 'prop-types'

import get from 'lodash/get'

import { models } from 'cozy-client'
import withBreakpoints from 'cozy-ui/transpiled/react/helpers/withBreakpoints'
import ListItem from 'cozy-ui/transpiled/react/MuiCozyTheme/ListItem'

import { fullContactPropTypes } from '../ContactPropTypes'
import withModal from '../HOCs/withModal'
import ContactCardModal from '../Modals/ContactCardModal'
import ContactWithSelection from './ContactSelection'
import ContactPhone from './Contacts/ContactPhone'
import ContactIdentity from './Contacts/ContactIdentity'
import ContactCozy from './Contacts/ContactCozy'
import ContactEmail from './Contacts/ContactEmail'

const { getPrimaryCozy, getPrimaryPhone, getPrimaryEmail } = models.contact

/**
 * Returns whether a document has been updated
 * @param {object} document - Initial document
 * @param {object} nextDocument - Potentially modified document
 * @returns {boolean}
 */
export const hasDocBeenUpdated = (document, nextDocument) =>
  document._rev !== nextDocument._rev ||
  get(document, 'cozyMetadata.updatedAt') !==
    get(nextDocument, 'cozyMetadata.updatedAt')

class ContactListItem extends Component {
  shouldComponentUpdate(nextProps) {
    const { contact, ...otherProps } = this.props

    const hasOtherPropsBeenChanged = Object.entries(otherProps).some(
      ([key, otherProp]) => {
        return otherProp !== nextProps[key]
      }
    )

    if (
      hasDocBeenUpdated(contact, nextProps.contact) ||
      hasOtherPropsBeenChanged
    ) {
      return true
    }
    return false
  }

  render() {
    const {
      contact,
      showModal,
      divider,
      breakpoints: { isMobile }
    } = this.props
    const email = getPrimaryEmail(contact) || undefined
    const phone = getPrimaryPhone(contact) || undefined
    const cozyUrl = getPrimaryCozy(contact) || undefined

    return (
      <ListItem
        className="u-c-pointer"
        data-testid="contact-listItem"
        divider={divider}
        onClick={() => showModal(<ContactCardModal id={contact._id} />)}
      >
        <ContactWithSelection contact={contact} />
        <ContactIdentity contact={contact} />
        {!isMobile && <ContactEmail email={email} />}
        {!isMobile && <ContactPhone phone={phone} />}
        {!isMobile && <ContactCozy cozyUrl={cozyUrl} />}
      </ListItem>
    )
  }
}

ContactListItem.propTypes = {
  contact: fullContactPropTypes.isRequired,
  showModal: PropTypes.func.isRequired,
  divider: PropTypes.bool
}
ContactListItem.defaultProps = {
  selection: null
}

export default withBreakpoints()(withModal(ContactListItem))
