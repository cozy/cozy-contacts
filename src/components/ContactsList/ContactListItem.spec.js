import React from 'react'
import renderer from 'react-test-renderer'
import { mount } from 'enzyme'

import AppLike from '../../tests/Applike'
import ContactListItem, { hasDocBeenUpdated } from './ContactListItem'

describe('ContactListItem', () => {
  it('should accept the strict minimum', () => {
    const contact = { email: [{ address: 'johndoe@localhost' }] }
    const contactListItemInstance = (
      <AppLike>
        <ContactListItem contact={contact} />
      </AppLike>
    )
    const contactListItem = mount(contactListItemInstance)
    const contactListItemEmail = contactListItem.find('ContactEmail')
    expect(contactListItemEmail).toBeDefined()
    expect(contactListItemEmail.text()).toBe(contact.email[0].address)
  })

  it('should display data', () => {
    const contact = {
      name: { familyName: 'Doe', givenName: 'John' },
      phone: [{ number: '0123456789' }],
      email: [{ address: 'johndoe@localhost' }],
      cozy: [{ url: 'http://johndoe.mycozy.cloud' }]
    }
    const ContactListItemInstance = (
      <AppLike>
        <ContactListItem contact={contact} />
      </AppLike>
    )
    const contactListItem = mount(ContactListItemInstance)
    const contactListItemName = contactListItem.find('ContactName')
    expect(contactListItemName).toBeDefined()
    expect(contactListItemName.text()).toEqual(
      expect.stringContaining(contact.name.givenName)
    )
    expect(contactListItemName.text()).toEqual(
      expect.stringContaining(contact.name.familyName)
    )
    const contactListItemPhone = contactListItem.find('ContactPhone')
    expect(contactListItemPhone).toBeDefined()
    expect(contactListItemPhone.text()).toBe(contact.phone[0].number)

    const contactListItemCozyurl = contactListItem.find('ContactCozy')
    expect(contactListItemCozyurl).toBeDefined()
    expect(contactListItemCozyurl.text()).toBe(contact.cozy[0].url)
  })

  it('should display default value for missing information', () => {
    const contact = {}
    const ContactListItemInstance = (
      <AppLike>
        <ContactListItem contact={contact} />
      </AppLike>
    )
    const contactListItem = mount(ContactListItemInstance)
    const contactListItemName = contactListItem.find('ContactName')
    expect(contactListItemName).toBeDefined()
    expect(contactListItemName.text().trim()).toBe('')
    const contactListItemPhone = contactListItem.find('ContactPhone')
    expect(contactListItemPhone).toBeDefined()
    expect(contactListItemPhone.text().trim()).toBe('—')
    const contactListItemCozyurl = contactListItem.find('ContactCozy')
    expect(contactListItemCozyurl).toBeDefined()
    expect(contactListItemCozyurl.text().trim()).toBe('—')
  })

  it('should accept empty array', () => {
    const contact = { email: [] }
    const ContactListItemInstance = (
      <AppLike>
        <ContactListItem contact={contact} />
      </AppLike>
    )
    const contactListItem = mount(ContactListItemInstance)
    const contactListItemEmail = contactListItem.find('ContactEmail')
    expect(contactListItemEmail).toBeDefined()
    expect(contactListItemEmail.text()).toBe('—')
  })

  it('should match the contact snapshot', () => {
    const contact = {
      name: { familyName: 'Doe', givenName: 'John' },
      phone: [{ number: '0123456789' }],
      email: [{ address: 'johndoe@localhost' }],
      cozy: [{ url: 'http://johndoe.mycozy.cloud' }]
    }
    const tree = renderer
      .create(
        <AppLike>
          <ContactListItem contact={contact} />
        </AppLike>
      )
      .toJSON()
    expect(tree).toMatchSnapshot()
  })
})

describe('hasDocBeenUpdated', () => {
  it('should return false if nothing changed', () => {
    const document = {
      contact: {
        name: { familyName: 'Doe', givenName: 'John' },
        _rev: '001',
        cozyMetadata: { updatedAt: '2020' }
      }
    }
    const nextDocument = {
      contact: {
        name: { familyName: 'Doe', givenName: 'John' },
        _rev: '001',
        cozyMetadata: { updatedAt: '2020' }
      }
    }
    const isUpdated = hasDocBeenUpdated(document, nextDocument)
    expect(isUpdated).toBe(false)
  })

  it('should return true if document rev changed', () => {
    const document = {
      name: { familyName: 'Doe', givenName: 'John' },
      _rev: '001',
      cozyMetadata: { updatedAt: '2020' }
    }
    const nextDocument = {
      name: { familyName: 'Doe', givenName: 'John' },
      _rev: '002',
      cozyMetadata: { updatedAt: '2020' }
    }
    const isUpdated = hasDocBeenUpdated(document, nextDocument)
    expect(isUpdated).toBe(true)
  })

  it('should return true if document updatedAt changed', () => {
    const document = {
      name: { familyName: 'Doe', givenName: 'John' },
      _rev: '001',
      cozyMetadata: { updatedAt: '2019' }
    }
    const nextDocument = {
      name: { familyName: 'Doe', givenName: 'John' },
      _rev: '001',
      cozyMetadata: { updatedAt: '2020' }
    }
    const isUpdated = hasDocBeenUpdated(document, nextDocument)
    expect(isUpdated).toBe(true)
  })

  it('should return true if document updatedAt and rev changed', () => {
    const document = {
      name: { familyName: 'Doe', givenName: 'John' },
      _rev: '001',
      cozyMetadata: { updatedAt: '2019' }
    }
    const nextDocument = {
      name: { familyName: 'Doe', givenName: 'John' },
      _rev: '002',
      cozyMetadata: { updatedAt: '2020' }
    }
    const isUpdated = hasDocBeenUpdated(document, nextDocument)
    expect(isUpdated).toBe(true)
  })

  it('should return true if document cozyMetadata is missing', () => {
    const document = {
      name: { familyName: 'Doe', givenName: 'John' },
      _rev: '001'
    }
    const nextDocument = {
      name: { familyName: 'Doe', givenName: 'John' },
      _rev: '001',
      cozyMetadata: { updatedAt: '2020' }
    }
    const isUpdated = hasDocBeenUpdated(document, nextDocument)
    expect(isUpdated).toBe(true)
  })
})
