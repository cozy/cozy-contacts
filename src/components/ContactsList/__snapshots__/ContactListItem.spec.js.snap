// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ContactListItem should match the contact snapshot 1`] = `
<li
  className="MuiListItem-root u-c-pointer MuiListItem-gutters"
  data-testid="contact-listItem"
  disabled={false}
  onClick={[Function]}
>
  <div
    className="contact-selection"
    onClick={[Function]}
  >
    <span
      data-input="checkbox"
    >
      <input
        checked={false}
        readOnly={true}
        type="checkbox"
      />
      <label />
    </span>
  </div>
  <div
    className="styles__TableCell___yJCq7 contact-identity u-flex u-flex-items-center u-ellipsis"
  >
    <div
      className="styles__c-avatar___PpDI- styles__c-avatar--text___2dvna"
      style={
        Object {
          "--circleSize": "32px",
          "backgroundColor": "var(--brightSun)",
        }
      }
    >
      
      <span
        className="styles__c-avatar-initials___310qC"
      >
        JD
      </span>
    </div>
    <p
      className="MuiTypography-root u-ml-1 MuiTypography-body1 MuiTypography-noWrap MuiTypography-gutterBottom"
      inline={true}
    >
      <span
        className=""
      >
        John
         
      </span>
      <span
        className="u-fw-bold"
      >
        Doe
         
      </span>
    </p>
  </div>
  <div
    className="styles__TableCell___yJCq7 contact-email u-ellipsis"
  >
    johndoe@localhost
  </div>
  <div
    className="styles__TableCell___yJCq7 contact-phone u-ellipsis"
  >
    0123456789
  </div>
  <div
    className="styles__TableCell___yJCq7 contact-cozyurl u-ellipsis"
  >
    http://johndoe.mycozy.cloud
  </div>
</li>
`;
